src/colony/modules/net.js

```diff
@@ -37,24 +43,88 @@ function isPipeName(s) {
 function toNumber(x) { return (x = Number(x)) >= 0 ? x : false; }
 
 /**
- * TCPSocket
+ * Socket
  */
 
-function TCPSocket (socket, _secure) {
-  Stream.Duplex.call(this);
+function Socket(opts) {
+  if (!(this instanceof Socket)) return new Socket(opts);
+  switch (typeof opts) {
+    case 'number':
+      opts = { fd: opts }; // Legacy interface.
+      break;
+    case 'undefined':
+      opts = {};
+      break;
+  }
+  stream.Duplex.call(this, opts);
   
-  if (typeof socket === 'object') {
-    this.socket = socket.fd;
-    // TODO: respect readable/writable flags
-    if (socket.allowHalfOpen) console.warn("Ignoring allowHalfOpen option.");
-  } 
+  this._opts = opts;
+  this._secure = opts._secure;
+  this._pendingCalls = [];
+}
+util.inherits(Socket, stream.Duplex);
+
+['_read', '_write', 'close', 'destroy', 'setTimeout', 'setNoDelay'].forEach(function (method) {
+  Socket.prototype[method] = function () {
+    this._pendingCalls.push({method:method, arguments:arguments});
+  };
+});
+
+Socket.prototype._doPending = function (proto) {
+  var self = this;
+  this._pendingCalls.forEach(function (call) {
+    proto[call.method].apply(self, call.arguments);
+  });
+  delete this._pendingCalls;
+};
+
+
+Socket.prototype.connect = function (opts, cb) {
+  // NOTE: imported here to avoid circular dependency
+  var net_proxied = require('_net_proxied');
+  
+  if (typeof opts !== 'object') {
+    var args = normalizeConnectArgs(arguments, this._secure);
+    return Socket.prototype.connect.apply(this, args);
+  }
+  
+  if (cb && this._secure) this.once('secureConnect', cb);
+  else if (cb) this.once('connect', cb);
+  
+  var self = this,
+      port = +opts.port,
+      host = opts.host || "127.0.0.1";
+  net_proxied._protoForConnection(host, port, this._opts, function (e, proto) {
+      if (e) return self.emit('error', e);
+      self.__proto__ = proto;   // HACK: convert to "concrete" subclass here, now that we know necessary type
+      self._setup(self._opts);  // pass original (constructor) opts
+      self._doPending(proto);
+      self._connect(port, host, cb);
+  });
+  
+  return this;
+};
 
-  this._secure = _secure;
+
+/**
+ * TCPSocket
+ */
+
+function TCPSocket (opts) {
+  if (!(this instanceof TCPSocket)) return new TCPSocket(opts);
```

Comment:
When it is called as a function `TCPSocket()` instead of a constructor `new TCPSocket()`.